File: sdks/packages/java/src/main/java/com/github/danielpaulus/goios/stream/SseReader.java
package com.github.danielpaulus.goios.stream;

import java.util.Iterator;
import java.util.NoSuchElementException;

/**
 * A pull-based Server-Sent Events reader over a line source.
 *
 * <p>Parses text/event-stream frames (blank-line delimited groups of
 * {@code event:} / {@code data:} lines, {@code :}-comment keep-alives ignored),
 * decodes each frame with the supplied {@link EventDecoder}, and yields typed
 * {@link SseEvent}s. Heartbeats are skipped unless {@code includeHeartbeats} is
 * set. Multi-line {@code data:} fields are joined with {@code \n}. A trailing
 * frame not terminated by a blank line is dispatched at end-of-stream.
 *
 * <p>Usable as an {@link Iterator} or in a {@code for-each} loop
 * ({@link Iterable}), and as an {@link AutoCloseable}: closing runs the
 * {@code onClose} hook exactly once (releasing the underlying HTTP response) and
 * stops further iteration, so a live stream can be cancelled mid-flight.
 */
public final class SseReader implements Iterator<SseEvent>, Iterable<SseEvent>, AutoCloseable {

    private final Iterator<String> lines;
    private final EventDecoder decoder;
    private final boolean includeHeartbeats;
    private final Runnable onClose;

    private SseEvent next;
    private boolean closed;
    private boolean onCloseRan;
    private boolean sawAnyLine; // whether we've consumed at least one line (for trailing-frame flush)

    public SseReader(Iterator<String> lines, EventDecoder decoder,
                     boolean includeHeartbeats, Runnable onClose) {
        this.lines = lines;
        this.decoder = decoder;
        this.includeHeartbeats = includeHeartbeats;
        this.onClose = onClose;
    }

    @Override
    public Iterator<SseEvent> iterator() {
        return this;
    }

    @Override
    public boolean hasNext() {
        if (next != null) {
            return true;
        }
        if (closed) {
            return false;
        }
        next = advance();
        return next != null;
    }

    @Override
    public SseEvent next() {
        if (!hasNext()) {
            throw new NoSuchElementException();
        }
        SseEvent ev = next;
        next = null;
        return ev;
    }

    /** Parse and decode the next dispatchable event, or {@code null} at end/close. */
    private SseEvent advance() {
        String eventName = null;
        StringBuilder data = null;
        boolean inFrame = false;

        while (!closed) {
            if (!lines.hasNext()) {
                // End of stream: flush a trailing frame with data but no blank terminator.
                if (inFrame && data != null) {
                    SseEvent ev = dispatch(eventName, data.toString());
                    return ev != null ? ev : null;
                }
                return null;
            }
            String line = lines.next();
            sawAnyLine = true;

            if (line.isEmpty()) {
                if (inFrame) {
                    SseEvent ev = dispatch(eventName, data == null ? "" : data.toString());
                    if (ev != null) {
                        return ev;
                    }
                    // Skipped (e.g. heartbeat): reset and keep scanning.
                    eventName = null;
                    data = null;
                    inFrame = false;
                }
                continue;
            }
            if (line.charAt(0) == ':') {
                // Comment / keep-alive line — ignore.
                continue;
            }
            inFrame = true;
            if (line.startsWith("event:")) {
                eventName = strip(line.substring("event:".length()));
            } else if (line.startsWith("data:")) {
                String chunk = strip(line.substring("data:".length()));
                if (data == null) {
                    data = new StringBuilder(chunk);
                } else {
                    data.append('\n').append(chunk);
                }
            }
            // Other field names (id:, retry:) are ignored.
        }
        return null;
    }

    /** Decode one frame, honoring heartbeat filtering; returns null if filtered out. */
    private SseEvent dispatch(String eventName, String data) {
        String name = eventName == null ? "message" : eventName;
        SseEvent ev = decoder.apply(name, data);
        if (ev instanceof HeartbeatEvent && !includeHeartbeats) {
            return null;
        }
        return ev;
    }

    private static String strip(String s) {
        // A single optional leading space after the colon is part of the SSE format.
        return s.startsWith(" ") ? s.substring(1) : s;
    }

    @Override
    public void close() {
        closed = true;
        next = null;
        if (!onCloseRan) {
            onCloseRan = true;
            if (onClose != null) {
                onClose.run();
            }
        }
    }
}

File: sdks/packages/csharp/src/GoIos.Sdk/SseReader.cs
using System.Runtime.CompilerServices;
using System.Text;
using System.Text.Json;

namespace GoIos.Sdk;

/// <summary>
/// Maps an SSE <c>event:</c> name plus its raw <c>data:</c> JSON to a typed
/// <see cref="SseEvent"/>. Return <c>null</c> to fall through to the reader's
/// unknown-event handling.
/// </summary>
public delegate SseEvent? SseEventFactory(string eventName, string data);

/// <summary>
/// A reusable, allocation-light parser for the go-ios Server-Sent Events wire
/// contract. Each frame is:
/// <code>
/// event: &lt;name&gt;\n
/// data: &lt;compact-json&gt;\n
/// \n
/// </code>
/// The reader tolerates split reads (a frame arriving across multiple chunks),
/// multiple frames per chunk, CRLF or LF line endings, and multi-line
/// <c>data:</c> fields (joined with <c>\n</c> per the SSE spec). Comment lines
/// (starting with <c>:</c>) are ignored.
/// </summary>
public static class SseReader
{
    /// <summary>
    /// Reads an event stream and yields typed events. Heartbeats are surfaced as
    /// <see cref="HeartbeatEvent"/>; frames whose <c>event:</c> name the
    /// <paramref name="factory"/> does not recognize are surfaced as
    /// <see cref="UnknownEvent"/>.
    /// </summary>
    public static async IAsyncEnumerable<SseEvent> ReadAsync(
        Stream stream,
        SseEventFactory factory,
        [EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        using var reader = new StreamReader(stream, Encoding.UTF8, detectEncodingFromByteOrderMarks: false);

        string? eventName = null;
        var data = new StringBuilder();
        bool sawData = false;

        while (!cancellationToken.IsCancellationRequested)
        {
            string? line;
            try
            {
                line = await reader.ReadLineAsync(cancellationToken).ConfigureAwait(false);
            }
            catch (OperationCanceledException)
            {
                yield break;
            }

            if (line is null)
            {
                // End of stream. Flush a trailing frame that had no blank-line terminator.
                if (eventName is not null || sawData)
                {
                    var evt = Dispatch(factory, eventName, data.ToString());
                    if (evt is not null) yield return evt;
                }
                yield break;
            }

            if (line.Length == 0)
            {
                // Blank line: dispatch the accumulated frame (if any).
                if (eventName is not null || sawData)
                {
                    var evt = Dispatch(factory, eventName, data.ToString());
                    if (evt is not null) yield return evt;
                }
                eventName = null;
                data.Clear();
                sawData = false;
                continue;
            }

            if (line[0] == ':')
            {
                // SSE comment / keep-alive colon line — ignore.
                continue;
            }

            var colon = line.IndexOf(':');
            string field, value;
            if (colon < 0)
            {
                field = line;
                value = "";
            }
            else
            {
                field = line.Substring(0, colon);
                value = line.Substring(colon + 1);
                if (value.StartsWith(' ')) value = value.Substring(1); // strip one leading space
            }

            switch (field)
            {
                case "event":
                    eventName = value;
                    break;
                case "data":
                    if (sawData) data.Append('\n');
                    data.Append(value);
                    sawData = true;
                    break;
                // "id" and "retry" are part of SSE but unused by this contract.
                default:
                    break;
            }
        }
    }

    private static SseEvent? Dispatch(SseEventFactory factory, string? eventName, string data)
    {
        var name = eventName ?? "message";

        if (name == "heartbeat")
            return new HeartbeatEvent { EventName = name };

        var typed = factory(name, data);
        if (typed is not null)
            return typed with { EventName = name };

        return new UnknownEvent { EventName = name, RawData = data };
    }

    /// <summary>
    /// Helper for factories: deserialize <paramref name="data"/> to
    /// <typeparamref name="T"/> using the SDK's JSON options, tolerating empty
    /// payloads (returns a default-constructed instance).
    /// </summary>
    public static T Deserialize<T>(string data) where T : new()
    {
        if (string.IsNullOrWhiteSpace(data))
            return new T();
        return JsonSerializer.Deserialize<T>(data, JsonOptions.Default) ?? new T();
    }
}

/// <summary>Shared System.Text.Json options for SSE payload deserialization.</summary>
internal static class JsonOptions
{
    public static readonly JsonSerializerOptions Default = new()
    {
        PropertyNameCaseInsensitive = true,
        NumberHandling = System.Text.Json.Serialization.JsonNumberHandling.AllowReadingFromString,
    };
}
